### DISCUSSION TITLE
Error using tbl() with snowflake connection

### DISCUSSION THREAD

**Discussion by PHilden-RP** [posit-dev/positron#4436] [Feedback]
I can successfully make a connection to my snowflake database, but the `tbl()` function does not appear to work on this, I get the following error message

no applicable method for 'tbl' applied to an object of class "Snowflake"

The same code works just fine in RStudio.

Overall loving Positron, thanks for your work!

**Comment by juliasilge** [comment]
This looks related to #3893 where someone else saw:

> no applicable method for 'tbl' applied to an object of class "Oracle"

We don't have a great answer for why that person's problems were resolved unfortunately. Can you share what version of Positron you are running?

I expect that you can run this just fine?

```r
library(DBI)
library(duckdb)
library(dbplyr)
library(dplyr)

con <- dbConnect(duckdb::duckdb())

dbWriteTable(con, "cars", mtcars)

tbl(con, "cars")
```

Can you share the output for `dbGetInfo()` for Snowflake from both RStudio and Positron, to see if there are any differences?

**Reply by PHilden-RP** [reply]
Hi Julia,

Thanks for the response! I can confirm I can run the above without issue.


in Positron dbGetInfo() returns the following

> dbGetInfo(con)
$dbname
[1] "HEALTHVERITY"

$dbms.name
[1] "Snowflake"

$db.version
[1] "8.31.1"

$username
[1] "philden"

$host
[1] ""

$port
[1] ""

$sourcename
[1] ""

$servername
[1] "Snowflake"

$drivername
[1] "Snowflake - Latest version supported by Snowflake: 2.21.1"

$odbc.version
[1] "03.80.0000"

$driver.version
[1] "3.3.1"

$odbcdriver.version
[1] "03.80"

$supports.transactions
[1] TRUE

$supports.catalogs
[1] TRUE

$supports.schema
[1] TRUE

$getdata.extensions.any_column
[1] TRUE

$getdata.extensions.any_order
[1] TRUE



and in RStudio

$dbname
[1] "HEALTHVERITY"

$dbms.name
[1] "Snowflake"

$db.version
[1] "8.31.1"

$username
[1] "philden"

$host
[1] ""

$port
[1] ""

$sourcename
[1] ""

$servername
[1] "Snowflake"

$drivername
[1] "Snowflake"

$odbc.version
[1] "03.80.0000"

$driver.version
[1] "3.3.1"

$odbcdriver.version
[1] "03.80"

$supports.transactions
[1] TRUE

$supports.catalogs
[1] TRUE

$supports.schema
[1] TRUE

$getdata.extensions.any_column
[1] TRUE

$getdata.extensions.any_order
[1] TRUE




I am running this version of Positron

Positron Version: 2024.08.0 (system setup) build 48
Code - OSS Version: 1.91.0
Commit: ed616b36655fb24d116108bdd833f1321704315b
Date: 2024-08-19T03:30:12.447Z
Electron: 29.4.0
Chromium: 122.0.6261.156
Node.js: 20.9.0
V8: 12.2.281.27-electron.0
OS: Windows_NT x64 10.0.22631


Also, interestingly in Positron I can knit the document to an html_notebook without issue or errors and get the expected results.

**Comment by simonpcouch** [comment]
Thanks for the additional information and calling out that you can render notebooks without issue! Those are both super helpful. Your driver and DB versions are all quite recent so that doesn't seem like an issue. I'm unfortunately not able to replicate on Positron 2024.08.0 (system setup) build 48, Windows 11 Pro quite yet.

Could you share your odbc, dplyr, and dbplyr package versions? Also, could you try with both the current CRAN odbc (i.e. with `pak::pak("odbc")` and then restarting R) and current dev (`pak::pak("r-dbi/odbc")` and restarting R)?

Also, could you share the output of the following R code 1) from RStudio, 2) from Positron, and 3) From within an `html_notebook` rendered with Positron?

``` r
Sys.getenv("PATH")
```


Note to future self--this is the code I used to connect and test:

``` r
library(odbc)
library(dplyr)
library(dbplyr)

con <- dbConnect(snowflake(), authenticator = "externalbrowser", UID = "my_email")
dbExecute(con, "USE WAREHOUSE DEFAULT_WH")
tbl(con, in_catalog("SNOWFLAKE", "INFORMATION_SCHEMA", "APPLICABLE_ROLES"))
```

**Comment by PHilden-RP** [comment]
Hi Simon,

Below are the current package versions requested

odbc 1.5.0
dplyr 1.1.4
dbplyr 2.5.0


I tried with the current odbc package as well as updating to the DEV version noted and the results were the same.


 Sys.getenv("PATH") ouput


From RStudio


"C:\\rtools44\\x86_64-w64-mingw32.static.posix\\bin;C:\\rtools44\\usr\\bin;C:\\Users\\philden\\AppData\\Local\\Programs\\R\\R-4.4.0\\bin\\x64;C:\\WINDOWS\\system32;C:\\WINDOWS;C:\\WINDOWS\\System32\\Wbem;C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0\\;C:\\WINDOWS\\System32\\OpenSSH\\;C:\\Program Files\\Positron\\bin;C:\\Users\\philden\\AppData\\Local\\Microsoft\\WindowsApps;C:\\Users\\philden\\AppData\\Local\\GitHubDesktop\\bin;C:\\Program Files\\RStudio\\resources\\app\\bin\\quarto\\bin;C:\\Program Files\\RStudio\\resources\\app\\bin\\postback"


From Positron

"C:\\rtools44/x86_64-w64-mingw32.static.posix/bin;C:\\rtools44/usr/bin;;c:\\Program Files\\Positron\\resources\\app\\quarto\\bin;C:\\WINDOWS\\system32;C:\\WINDOWS;C:\\WINDOWS\\System32\\Wbem;C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0\\;C:\\WINDOWS\\System32\\OpenSSH\\;C:\\Program Files\\Positron\\bin;C:\\Users\\philden\\AppData\\Local\\Microsoft\\WindowsApps;C:\\Users\\philden\\AppData\\Local\\GitHubDesktop\\bin"


From html_notebook output

"C:\\rtools44/x86_64-w64-mingw32.static.posix/bin;C:\\rtools44/usr/bin;;c:\\Program Files\\Positron\\resources\\app\\quarto\\bin;C:\\WINDOWS\\system32;C:\\WINDOWS;C:\\WINDOWS\\System32\\Wbem;C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0\\;C:\\WINDOWS\\System32\\OpenSSH\\;C:\\Program Files\\Positron\\bin;C:\\Users\\philden\\AppData\\Local\\Microsoft\\WindowsApps;C:\\Users\\philden\\AppData\\Local\\GitHubDesktop\\bin"


Best,


Pat

________________________________

**Comment by simonpcouch** [comment]
Looks like `PATH`s are not the issue.

Could you supply the code snippet you're using to connect? All is well if you scrub out usernames/passwords/etc.

Only one more hypothesis I can think of: I don't know how Positron's connections contract is implemented but perhaps RStudio's attaches a namespace somewhere where Positron doesn't. Have you explicitly loaded each of DBI, odbc, dplyr, and dbplyr (rather than using e.g. `dplyr::tbl()`? Are results the same when you load them all and don't namespace vs. load none and refer to them all by their namespace?

**Comment by PHilden-RP** [comment]
Hi Simon,

I was able to resolve the issue by explicitly loading the odbc package.

Previously, my connection was done by

con <- dbConnect(
  odbc::odbc(), 
  ....
)

as that was the only time I was calling the odbc package.

When I replace the above with

library(odbc)

con <- dbConnect(
  odbc(),
  ...
  )

Everything works without issue and I no longer have the error with the tbl() function.

All best,


Pat

________________________________

**Comment by simonpcouch** [comment]
Awesome, thanks for confirming!

**Comment by juliasilge** [comment]
We're going to keep an eye on this moving forward, as we've now had two people run into problems connecting to databases with S3 method errors like:

> no applicable method for 'tbl' applied to an object of class "Foo"

I am particularly suspicious that something may be wrong in Positron because @PHilden-RP said his code without the explicit `library(odbc)` was able to run via knitr but not in the console. 

Thank you again for the report!
